Document UDT improvements and breaking changes [HZ-3497] [HZ-3686] (#969)
hazelcast/hz-docs: docs/modules/sql/pages/user-defined-types.adoc

    public Long id;
    public String name;
    public Office office;
}

class Office implements Serializable {
    public Long id;
    public String name;
}
----

NOTE: The name of a type can differ from the one specified in the Java class or Portable/Compact/Avro schema. However, types must have distinct names within the set of names across all mappings and views as they share the same namespace.

[#organization-office-types]
[source,sql]
----
CREATE TYPE Organization (
    id BIGINT
    name VARCHAR,
    office Office
) OPTIONS (
    'javaClass'='com.example.Organization'
);

CREATE TYPE Office (
    id BIGINT
    name VARCHAR
) OPTIONS (
    'javaClass'='com.example.Office'
);
----

== Creating Mappings

NOTE: The `organization` column is explicitly specified as `Organization` to prevent it being auto-resolved as a generic `OBJECT`, and therefore unable to query its sub-columns.

[#users-mapping]
[source,sql]
----
CREATE MAPPING users (
    __key BIGINT,
    id BIGINT,
    name VARCHAR,
    organization Organization
) TYPE IMap OPTIONS (
    'keyFormat'='bigint',
    'valueFormat'='java',
    'valueJavaClass'='com.example.User'
);
----

== Support for Cycles

> When creating a UDT, the existence of referenced types are only verified when the type is used in a `CREATE MAPPING` statement. This makes it possible to create cyclic types.

Reviewer: ```suggestion
When creating a UDT, the existence of referenced types is only verified when the type is used in a `CREATE MAPPING` statement. This makes it possible to create cyclic types.
```

= User-Defined Types in SQL

User-Defined Types (also known as UDTs, nested types, or nested fields) allow you to create custom data types that can be referenced in the xref:sql:create-mapping.adoc[CREATE MAPPING statement]. UDTs are also useful for creating and accessing hierarchical data structures.

== Feature Overview

> - UDTs can only be incorporated into mappings that has `portable`, `compact`, `java` or `avro` format although UDTs themselves are not tied to a specific format.

Author: 1. What is wrong with "incorporated into"?
2. "a ... format" should be changed to "the ... format".

> - UDTs can only be incorporated into mappings that has `portable`, `compact`, `java` or `avro` format although UDTs themselves are not tied to a specific format.

Reviewer 2: 1. I didn't say into was wrong. For me, incorporate = add to a body, so the "to" is already present. If you feel strongly, leave it as into.
2.  As you give alternatives, "a" is more appropriate, but "the" is not going to be wrong. It was the lack of an article (definite or indefinite) that I was commenting on.

> - UDTs can only be incorporated into mappings that has `portable`, `compact`, `java` or `avro` format although UDTs themselves are not tied to a specific format.

Author: 90767538

- Type options override mapping options. Where both the mapping and type define a schema (Portable class definition, Compact schema, Java class or Avro schema) and the corresponding mapping field is `__key` or `this`, the schema defined by the type is used. This means that the mapping does not need to define a schema if the type defines one.
- If the type does not specify a schema, it is resolved from the parent structure, which may be a mapping or another type.

> - Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a `CREATE MAPPING` statement. You can check the resolved fields using a `GET_DDL` query. The field resolution feature has the following limitations.

Author: I can accept this, but in English, sentences cannot end with a semicolon. The following usage is acceptable, which forms a single sentence.
> Sentence begins:
    1. words,
    2. words, and
    3. words.

> - Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a `CREATE MAPPING` statement. You can check the resolved fields using a `GET_DDL` query. The field resolution feature has the following limitations.

Reviewer 2: It isn't a semicolon, it is a colon. 

The colon is used because the sentence is a lead-in to the list.

> - Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a `CREATE MAPPING` statement. You can check the resolved fields using a `GET_DDL` query. The field resolution feature has the following limitations.

Author: 90767538

= User-Defined Types in SQL

User-Defined Types (also known as UDTs, nested types, or nested fields) allow you to create custom data types that can be referenced in the xref:sql:create-mapping.adoc[CREATE MAPPING statement]. UDTs are also useful for creating and accessing hierarchical data structures.

== Feature Overview

- `portable`, `compact`, `java` and `avro` formats are supported.

> - Cyclic definitions are only supported for Java-serialized structures. Portable and Compact formats do not support cyclic schemas, and Avro support is currently limited to acyclic schemas.

Reviewer: I don't think cyclic stuctures are a first thing that comes to mind when someone reads about UDTs. We can keep it in "Feature Overview" but move it to be last in the list.

Author: fef8c8c2

+

> NOTE: If mappings use cyclic structures, the `INSERT` and `UPDATE` queries are disabled.

Reviewer: this is repeated later in "Support for Cycles". IMO this is too much detail for "feature overview". In this section we can mention that we support cycles but there are important limitations described in detail later.

Author: fef8c8c2

> - Type options override mapping options. Where both the mapping and type define a schema (Portable class definition, Compact schema, Java class or Avro schema) and the corresponding mapping field is `__key` or `this`, the schema defined by the type is used. This means that the mapping does not need to define a schema if the type defines one.

Reviewer: mapping -> MAPPING? we mean specifically SQL MAPPING not any mapping

Author: Are you suggesting capitalizing all "mapping" usages? The name of the page is "User-Defined Types in SQL" and I think it is not ambiguous what "mapping" refers to.

Reviewer: we can keep this as is

- If the type does not specify a schema, it is resolved from the parent structure, which may be a mapping or another type.
- Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a `CREATE MAPPING` statement. You can check the resolved fields using a `GET_DDL` query. The field resolution feature has the following limitations.
 ** For Portable, Compact, or Avro formats, schema fields cannot be complex (`PORTABLE`, `COMPACT`, or `RECORD`). Otherwise, an exception will be thrown once the type is used in a mapping.
 ** For Portable format, portable IDs (`portableFactoryId`, `portableClassId` and `portableVersion`) must be unique within the cluster. Otherwise, deserialization issues can occur.
 ** For Portable format, the field resolution relies on the Portable class being already registered (by internal means, such as adding a configuration for Portable serialization) on the member executing the `CREATE MAPPING` command.
 ** For Java format, complex class fields are allowed and mapped to `OBJECT`.

>  ** For Java format, only the fields declared by the type class are considered. The fields inherited from superclasses are ignored.

Reviewer: this is a bit surprising? were inherited fields ignored before?

Reviewer: ok, at least in the docs it was said that it is not supported

== Creating Types

To create a new type, use the `CREATE TYPE` statement:

[source,sql]
----
CREATE [OR REPLACE] TYPE [IF NOT EXISTS] MyTypeName [(
    colName colType,
    ...
)] [OPTIONS (
    'javaClass'='com.mypackage.MyJavaClass' <1>
  | 'compactTypeName'='MyCompactRecordName' <2>
  | 'portableFactoryId'='123', 'portableClassId'='456', ['portableVersion'='789'] <3>
  | 'avroSchema'='{"type":"record","name":"myType","fields":[{"name":"colName","type":"colType"},...]}' <4>
)]
----
<1> In `java` format, you can use the `javaClass` option to override the type class with a subclass. This is necessary if the original type class is abstract or interface, and you want to use `INSERT` and `UPDATE` statements.

> <2> In `portable` format, you can use the `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type. This is only effective when the type is used for `__key` or `this` fields. `portableVersion` defaults to 0 if not specified. Portable ID components must be defined together; incomplete definitions have the same effect as no definition.

Reviewer: ```suggestion
<2> In `portable` format, you can use the `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type. This is only effective when the type is used for `__key` or `this` fields. `portableVersion` defaults to 0 if not specified. Required Portable ID components must be defined together; incomplete definitions are ignored.
```

Author: fef8c8c2

<4> In `avro` format, you can use the `avroSchema` option to specify an inline Avro schema. This is only effective when the type is used for `__key` or `this` fields.

[NOTE]
====
. `EXTERNAL NAME` aliases are not supported for UDTs; column names must have the same name as their corresponding Java class field or Portable/Compact/Avro schema field.
. You can mix options that belong to different formats. When you create a mapping that references your UDT, the relevant options are used and the others are ignored, which makes it possible to use a UDT in multiple mappings having different formats.
====

=== Replacing Types and Type Consistency
Currently, there is a limitation on the replacement of existing types:
if the replaced type was already used in a mapping, you need to recreate that mapping
to update its data type information using the `DROP MAPPING` and `CREATE MAPPING` statements.
However, if the type hierarchy was not used in a mapping, any type in that hierarchy can be safely
changed, and these changes will appear in the new mapping. This is because the links

> between types are symbolic (based on the name only), and they are materialized only when used in a mapping.

Reviewer: and fields for fieldless UDT are captured and stored at this time

Author: In Feature Overview, it says
> Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a `CREATE MAPPING` statement.

Thus, both the field resolution and materialization happens when the type is used in mapping. They are mentioned in different sections, which is, I think, not a shortcoming since the documentation is not pivoted around mapping/type lifecycle.

Reviewer: let's move discussion to https://github.com/hazelcast/hz-docs/pull/969#discussion_r1466279286

    public Long id;
    public String name;
    public Office office;
}

class Office implements Serializable {
    public Long id;
    public String name;
}
----

NOTE: The name of a type can differ from the one specified in the Java class or Portable/Compact/Avro schema. However, types must have distinct names within the set of names across all mappings and views as they share the same namespace.

[#organization-office-types]
[source,sql]
----
CREATE TYPE Organization (
    id BIGINT
    name VARCHAR,
    office Office
) OPTIONS (
    'javaClass'='com.example.Organization'
);

CREATE TYPE Office (
    id BIGINT
    name VARCHAR
) OPTIONS (
    'javaClass'='com.example.Office'
);
----

== Creating Mappings

NOTE: The `organization` column is explicitly specified as `Organization` to prevent it being auto-resolved as a generic `OBJECT`, and therefore unable to query its sub-columns.

[#users-mapping]
[source,sql]
----
CREATE MAPPING users (
    __key BIGINT,
    id BIGINT,
    name VARCHAR,
    organization Organization
) TYPE IMap OPTIONS (
    'keyFormat'='bigint',
    'valueFormat'='java',
    'valueJavaClass'='com.example.User'
);
----

== Support for Cycles

> Cycles between types are only supported for Java format. However, this is limited to querying, not inserting or updating.

Reviewer: ```suggestion
Cycles between types are only supported for Java format. However, the support is limited only to querying. Inserting or updating with cyclic types is currently not supported.
```

Author: fef8c8c2

> NOTE: The existence of referenced types are only verified when the type is used in a `CREATE MAPPING` statement.

Reviewer: ```suggestion
NOTE: It is possible to reference a not yet existing type (including the type being defined) in type definition which is necessary to create cyclic structure. The existence of referenced types is only verified when the type is used in a `CREATE MAPPING` statement.
```
Without the addition I did not understand how this note relates to cyclic types.

"including the type being defined" - is that true?

Author: Yes, it is true. (fef8c8c2)

=== Enabling Cycling Type Support

> You can enable cyclic types by setting the `hazelcast.sql.experimental.custom.cyclic.types.enabled` property to `true` in the member configuration.

Reviewer: ```suggestion
You can enable cyclic types by setting the `hazelcast.sql.experimental.custom.cyclic.types.enabled` property to `true` in the member configuration. It is disabled by default.
```

Author: fef8c8c2

= User-Defined Types in SQL

User-Defined Types (also known as UDTs, nested types, or nested fields) allows you to create custom data types that can be referenced in the xref:sql:create-mapping.adoc[CREATE MAPPING statement].

UDTs are useful for creating and accessing hierarchical data structures, which can even be cyclic. See <<queryingExamples,examples>> for more information.

== Feature Overview

- All serialization formats (namely `portable`, `compact`, `java` and `avro`) are supported.
- Cyclic definitions are only supported for Java-serialized structures. (Portable and Compact formats do not support cyclic schemas at all. Avro support is currently limited to acyclic schemas.) Note that `INSERT`/`UPDATE` queries are disabled for mappings that use cyclic structures.
- Type options _always_ override mapping options. Specifically, if the mapping has only `__key` or `this` fields, and both the mapping and type define a schema (Portable class definition, Compact schema, Java class or Avro schema), the one defined by the type is used. This also means that, in such situations, as long as the type defines a schema, the mapping does not need to.

> - If type schema is not specified, it will be resolved from the parent structure, which may be a mapping or another type.

Author: "the parent structure must provide the schema in a mapping or another type" should be "it will be resolved from the parent structure, which may be a mapping or another type". The parent structure **is** a mapping or type; it cannot specify schema in a mapping or type.

> - If type schema is not specified, it will be resolved from the parent structure, which may be a mapping or another type.

Reviewer 2: Amended suggestion

> - Type fields are optional and will be resolved from the schema if they are not specified. In that case, schema fields cannot be complex (`PORTABLE`/`COMPACT`/`RECORD`) for Portable/Compact/Avro formats. For Java format, complex fields are allowed and mapped to `OBJECT`. Type fields are resolved and materialized once the type is used in a `CREATE MAPPING` statement. The resolved fields can be checked by issuing a `GET_DDL` query after using the type in a mapping.

Author: 1. When using Portable, Compact, or Avro formats, schema fields **can** be complex. You deleted "In that case", which is why the meaning changed.
2. "If using" sounds strange to me, "when using" might be better.
3. "`OBJECT`s" should be "`OBJECT`".

> - Type fields are optional and will be resolved from the schema if they are not specified. In that case, schema fields cannot be complex (`PORTABLE`/`COMPACT`/`RECORD`) for Portable/Compact/Avro formats. For Java format, complex fields are allowed and mapped to `OBJECT`. Type fields are resolved and materialized once the type is used in a `CREATE MAPPING` statement. The resolved fields can be checked by issuing a `GET_DDL` query after using the type in a mapping.

Reviewer 2: amended

= User-Defined Types in SQL

User-Defined Types (also known as UDTs, nested types, or nested fields) allow you to create custom data types that can be referenced in the xref:sql:create-mapping.adoc[CREATE MAPPING statement].

> UDTs are useful for creating and accessing hierarchical data structures, which can even be cyclic. See <<queryingExamples,examples>> for more information.

Reviewer: I would not mention cyclic types in into. They only work for Java format and require enabling feature flag

Author: 099f756d

== Feature Overview

> - All serialization formats (namely `portable`, `compact`, `java` and `avro`) are supported.

Reviewer: mapping format != serialization format. In particular Avro is not HZ serialization format, DS/IDS is not explicitly configured in SQL.

Reviewer: BTW, we don't support JSON for UDTs, do we? (it is supported as MAPPING format)

Author: I removed the "serialization" word. (099f756d) We don't support JSON for UDTs.

- Cyclic definitions are only supported for Java-serialized structures. Portable and Compact formats do not support cyclic schemas, and Avro support is currently limited to acyclic schemas.
+
NOTE: If mappings use cyclic structures, the `INSERT` and `UPDATE` queries are disabled.
- Type options override mapping options. Where both the mapping and type define a schema (Portable class definition, Compact schema, Java class or Avro schema) and the mapping contains only `__key` or `this` fields, the schema defined by the type is used. This means that the mapping does not need to define a schema if the type defines one.
- If the type does not specify a schema, the parent structure is used. The parent structure could be a mapping or another type.

> - Type fields are optional and are resolved from the schema if not specified. Following a `CREATE MAPPING` statement, type fields are resolved and materialized. You can check the resolved fields using a `GET_DDL` query after the type has been used in a mapping. If type fields are not specified:

Reviewer: AFAIK this "magic" materialization was a side effect and not something uses should rely on, as it is subject to change in the future

Reviewer: and were not sure if it makes sense in the long term

Author: I removed it from notes. Now, it is only mentioned in "Replacing Types and Type Consistency". (099f756d)

Reviewer: I do not see the place you mentioned. However I can see:

> If the type does not specify a schema, it is resolved from the parent structure, which may be a mapping or another type.

(and next bullet) which makes this magic UDT field resolution a "normal" feature, but we had doubts if we want to keep it

Author: Field materialization is mentioned in "Replacing Types and Type Consistency" as the following, and I think this is the right place for such an implementation detail.
> ... if the type hierarchy was not used in a mapping, any type in that hierarchy can be safely changed, and these changes will appear in the new mapping. **This is because the links between types are symbolic (based on the name only), and they are materialized only when used in a mapping.**

Reviewer: this sentence is about UDT references ("links between types") not fields of fieldless UDT. It is not obvious that is it the same.

And this is not only implementation detail. It has visible impact on the user. We say that:

> Type fields are optional and if not specified, they are resolved from the class/schema when the type is used in a CREATE MAPPING statement.

but we do not say what happens if the underlying schema changes (eg. Java class is updated).
On it's own the above sentence could mean that fieldless UDT may get different set of fields for each context in which it used which is not the way we work.

Author: 7bc63e25

 ** For Portable, Compact, or Avro formats, schema fields cannot be complex (`PORTABLE`, `COMPACT`, or `RECORD`
 ). 
 ** For Java format, complex fields are allowed and map to `OBJECT`. 

> == Enabling Cycling Type Support

Reviewer: I would move this section toward end of the document

Author: It is now mentioned in "Support for Cycles". (099f756d)

[source,sql]
----
CREATE [OR REPLACE] TYPE [IF NOT EXISTS] MyTypeName [(
    colName colType,
    ...
)] [OPTIONS (
    'javaClass'='com.mypackage.MyJavaClass' <1>
  | 'compactTypeName'='MyCompactRecordName' <2>
  | 'portableFactoryId'='123', 'portableClassId'='456', ['portableVersion'='789'] <3>
  | 'avroSchema'='{"type":"record","name":"myType","fields":[{"name":"colName","type":"colType"},...]}' <4>
)]
----

> <1> In `java` format, you can use `javaClass` option to override the type class with a subclass. Note that you must provide a concrete class for `INSERT` statement to work, which makes `javaClass` option mandatory when the original type class is abstract or interface.

Author: "You must provide a concrete class for `INSERT` statements" can be replaced with "You must provide a concrete class to be able to `INSERT`".

> <1> In `java` format, you can use `javaClass` option to override the type class with a subclass. Note that you must provide a concrete class for `INSERT` statement to work, which makes `javaClass` option mandatory when the original type class is abstract or interface.

Reviewer 2: No, it cannot. "to be able to" is not a good phrase.

How about "You must provide a concrete class to use `INSERT` statements" ?

> <1> In `java` format, you can use `javaClass` option to override the type class with a subclass. Note that you must provide a concrete class for `INSERT` statement to work, which makes `javaClass` option mandatory when the original type class is abstract or interface.

Reviewer 2: But then the sentence has no agreement: use INSERT statements vs use an INSERT statement

statements is the correct term here.

> <1> In `java` format, you can use `javaClass` option to override the type class with a subclass. Note that you must provide a concrete class for `INSERT` statement to work, which makes `javaClass` option mandatory when the original type class is abstract or interface.

Reviewer 2: **The INSERT statement** is a type of SQL query
**An INSERT statement** is an instance of the INSERT SQL query type
**INSERT statements** is a generalised way of referring to INSERT SQL queries

I may not be a programmer, but I have worked in software development for over 30 years.

I really appreciate the work you are putting into this, and the time you are taking to explain things (such as the tuples conversation). However, on this point, I remain unmovable.

How about: "You must provide a concrete class for each `INSERT` statement;", would that allow us to agree?

> <1> In `java` format, you can use `javaClass` option to override the type class with a subclass. Note that you must provide a concrete class for `INSERT` statement to work, which makes `javaClass` option mandatory when the original type class is abstract or interface.

Author: This is not really about software development. In English, a noun can refer to a "concept" instead of a particular instance (or instances). In such a case, it becomes uncountable, and consequently singular. For example, we can say that "police closed the investigation" (without a/the/-s) if we are not interested in a particular police. You must have seen such usages before in the news(papers). "`INSERT` statement" is not wrong (even without "the"), and it looks more natural to me. If you disagree, your suggestion can stay as it is.

> <1> In `java` format, you can use `javaClass` option to override the type class with a subclass. Note that you must provide a concrete class for `INSERT` statement to work, which makes `javaClass` option mandatory when the original type class is abstract or interface.

Reviewer 2: I disagree.

I understand, but, in this case `INSERT` = police, `INSERT` statement = the police. 

Police is a group term, or with your obvious mathematical bias, uncountable. Police is used when speaking of the force in general.
The police is specific, hence the definite article, used when we think the reader knows what we are referring to. 
The plural of police is still police.
You would never say a police; it would be a policeman, an officer, a policewoman.

This isn't a news(paper). It is technical communication. We need to be clear (hence why we don't use LaTex and mathematical equations, which "might be a bit confusing to the reader"). We need to make sure our information is clear and usable, with no misinterpretation, or cognitive dissonance.

You can check Microsoft's SQL docs to check the use if you like; for example: https://learn.microsoft.com/en-us/sql/t-sql/queries/top-transact-sql?view=sql-server-ver16

> <2> In `portable` format, you can use `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type, which is only effective when the type is used for `__key` or `this` fields. `portableVersion` is optional, similarly to `CREATE MAPPING` statement. However, it is not possible to inherit some components from the parent, so portable ID components must be defined together. Incomplete definitions have the same effect as no definition.

Author: "like the `CREATE MAPPING` statement" means that `CREATE MAPPING` statement is optional. It should be "like in" or "similarly to".

> <2> In `portable` format, you can use `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type, which is only effective when the type is used for `__key` or `this` fields. `portableVersion` is optional, similarly to `CREATE MAPPING` statement. However, it is not possible to inherit some components from the parent, so portable ID components must be defined together. Incomplete definitions have the same effect as no definition.

Reviewer 2: Why do we need the comparison? It is optional; what else does the user need to know? What information are you trying to communicate by the comparison?

> <2> In `portable` format, you can use `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type, which is only effective when the type is used for `__key` or `this` fields. `portableVersion` is optional, similarly to `CREATE MAPPING` statement. However, it is not possible to inherit some components from the parent, so portable ID components must be defined together. Incomplete definitions have the same effect as no definition.

Author: Saying `portableVersion` is optional might imply that it is only optional in the `CREATE TYPE` statement if the user hasn't read the `CREATE MAPPING` statement yet. I want them to know that `version` of a portable ID is always optional. However, you can drop that part if you think that it complicates the explanation unnecessarily.

> <2> In `portable` format, you can use `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type, which is only effective when the type is used for `__key` or `this` fields. `portableVersion` is optional, similarly to `CREATE MAPPING` statement. However, it is not possible to inherit some components from the parent, so portable ID components must be defined together. Incomplete definitions have the same effect as no definition.

Reviewer 2: Dropped the comparison

<4> In `avro` format, you can use `avroSchema` option to specify an inline Avro schema, which is only effective when the type is used for `__key` or `this` fields.

NOTE: `EXTERNAL NAME` aliases are not supported for types, column names must have the same name as their corresponding Java class field or Portable/Compact/Avro schema field.

NOTE: You can mix options that belong to different formats. When you create a mapping that references your UDT, the relevant options are used and the others are ignored, which makes it possible to use a UDT in multiple mappings having different serialization formats.

NOTE: To reference another type, you must provide the column list. Otherwise, in Java format, the column will be resolved as `OBJECT` type; in Portable, Compact and Avro formats, an exception will be thrown once the type is used in a mapping.

=== Java Format notes
For `java` format, if the column list is omitted, it will be automatically resolved from the corresponding
Java class. Note that the column list will only be extracted from the source class itself;
if it has columns that are inherited from a superclass, these columns will not be resolved.

=== Portable Format notes

> When using the `portable` format, make sure that (`factoryId`, `classId`, `version`) tuples are unique within the cluster. Otherwise, deserialization issues may happen.

Author: You shouldn't change the tuple notation as it has a mathematical meaning, which is very hard to express using plain English: (`factoryId`, `classId`, `version`) tuples. "mut" should be "must".

> When using the `portable` format, make sure that (`factoryId`, `classId`, `version`) tuples are unique within the cluster. Otherwise, deserialization issues may happen.

Reviewer 2: So, this is a mathematical tuple, not a Python tuple data type? A mathematical tuple is a finite sequence of elements, I believe, and the Python data type stores an ordered and unchangeable collection of data.

Are you listing available tuples, or providing an example of a tuple?

Either way, I feel that this needs more explanation, to effectively explain the issue and solution.

And, yes, mut ought to be must ;-) sorry, didn't mean to suggest a new typo!

> When using the `portable` format, make sure that (`factoryId`, `classId`, `version`) tuples are unique within the cluster. Otherwise, deserialization issues may happen.

Author: 1. "`factoryId`, `classId`, and `version` tuples" means that `factoryId`, `classId`, and `version` are tuples themselves.
2. "(`factoryId`, `classId`, `version`) tuples" means that we are talking about 3-tuples, whose elements denote `factoryId`, `classId` and `version`.

> When using the `portable` format, make sure that (`factoryId`, `classId`, `version`) tuples are unique within the cluster. Otherwise, deserialization issues may happen.

Reviewer 2: Amended the comment so that it, I hope, meets both our concerns

> When using the `portable` format, make sure that (`factoryId`, `classId`, `version`) tuples are unique within the cluster. Otherwise, deserialization issues may happen.

Author: "(`factoryId`, `classId`, and `version`)" doesn't conform the tuple notation and "triple" is unnecessary. If there is a way to render LaTeX, you can convert the tuple to a mathematical expression. Otherwise, you have to mix math and plain English. I admit that it might be a bit confusing to the reader, but non-standard notation doesn't help.

> When using the `portable` format, make sure that (`factoryId`, `classId`, `version`) tuples are unique within the cluster. Otherwise, deserialization issues may happen.

Author: Alternatively, you can say "that portable IDs (`factoryId`, `classId`, `version`) are unique".

  | 'avroSchema'='{"type":"record","name":"myType","fields":[{"name":"colName","type":"colType"},...]}' <4>
)]
----
<1> In `java` format, you can use the `javaClass` option to override the type class with a subclass. You must provide a concrete class for `INSERT` statements; this means that the `javaClass` option is mandatory when the original type class is abstract or interface.
<2> In `portable` format, you can use the `portableFactoryId`, `portableClassId` and `portableVersion` options to specify a portable ID for the type. This is only effective when the type is used for `__key` or `this` fields. `portableVersion` is optional. However, components cannot be inherited from the parent, so portable ID components must be defined together. Incomplete definitions have the same effect as no definition.
<3> In `compact` format, you can use the `compactTypeName` option to specify the Compact record name. If unspecified, it defaults to `<TypeName>CompactType`.
<4> In `avro` format, you can use the `avroSchema` option to specify an inline Avro schema. This is only effective when the type is used for `__key` or `this` fields.

> NOTE: `EXTERNAL NAME` aliases are not supported for types, column names must have the same name as their corresponding Java class field or Portable/Compact/Avro schema field.

Reviewer: ```suggestion
```

note is duplicated

[NOTE]
====
. `EXTERNAL_NAME` aliases  are not supported for types, column names must have the same name as their corresponding Java class field or Portable/Compact/Avro schema field.
. You can mix options that belong to different formats. When you create a mapping that references your UDT, the relevant options are used and the others are ignored, which makes it possible to use a UDT in multiple mappings having different serialization formats.
. To reference another type, you must provide the column list. Otherwise, in Java format, the column will be resolved as `OBJECT` type; in Portable, Compact and Avro formats, an exception will be thrown once the type is used in a mapping.
====

> NOTE: You can mix options that belong to different formats. When you create a mapping that references your UDT, the relevant options are used and the others are ignored, which makes it possible to use a UDT in multiple mappings having different serialization formats.

Reviewer: I think " it possible to use a UDT in multiple mappings having different serialization formats" is quite important feature (and significant improvement over previous approach) and should be mentioned in the overview. Do we have examples for that like read from kafka and write to IMap with the same UDT (OFC the syntax is not pretty as you have to recreate the object from fields)?

Author: Since there is no `format` option for UDTs, users would expect them to be format-agnostic. That's why, I think, we don't need to highlight this fact.

Reviewer: but we say in the beginning that "portable, compact, java and avro formats are supported" - by analogy with mappings it might be assumed that UDT has also single assigned format. IMO worth mentioning this explicitly.

Reviewer: this is documented explicitly only in note in "Creating Types"

Author: fef8c8c2

> NOTE: To reference another type, you must provide the column list. Otherwise, in Java format, the column will be resolved as `OBJECT` type; in Portable, Compact and Avro formats, an exception will be thrown once the type is used in a mapping.

Reviewer: we have more notes than normal text here: we have "notes" boxes and "java format notes" and "protable format notes" sections. IMO it would be simpler if we had normal section for each format

If a Type hierarchy contains cycles, any mapping using any of these types (provided that type is not itself an acyclic branch)
will have `INSERT` and `UPDATE` commands disabled.
Additionally, support for cycles also means no validation for existence of custom types at the time of `CREATE TYPE` execution.
**Type hierarchies are only verified for consistency upon actual use in `CREATE MAPPING`.**
Cycles between types are only supported for Java format; however, this is limited to queries. 

NOTE: The existence of referenced types are only verified when used in a `CREATE MAPPING` statement.

=== Replacing Types and Type Consistency
Currently, there is a limitation on the replacement of existing types:
if the replaced type was already used in a mapping, you need to fully replace that mapping
to update its data type information using the `DROP MAPPING` and `CREATE MAPPING` statements.
However, if the type hierarchy was not used in a mapping, any type in that hierarchy can be safely
changed, and these changes will appear in the new mapping. This is because the links
between types are symbolic (based on the name only), and they're only "materialized" once used in a mapping.

> === CREATE TYPE Examples

Author: 099f756d

}
----

=== Creating Types [[normalTypeDefinitions]]

> NOTE: Types can have a different name than their Java class or Portable/Compact/Avro schema. The only limitation is that the types must have distinct names within the set of names of all mappings and views as they all share the same namespace.

Author: I will accept this if you change
1. "names used in all mappings" to "names of all mappings", and
2. "a namespace" to "the same namespace".

== Querying Support

> Querying is provided with the field access operator which has the following syntax (parentheses are required):

Author: Parenthesis is singular.

> Querying is provided with the field access operator which has the following syntax (parentheses are required):

Reviewer 2: Great, make it plural then (I would have used brackets myself, to keep the level of English lower, but I was trying to keep your wording).

Typos are for everyone, you know ;-)

> `mappingColumn` must be the top-level column inside a mapping that has a UDT as its type,

Reviewer: I would add a note pointing out that parentheses are required. This is very surprising quirk of our syntax

Author: fef8c8c2

SELECT (organization).office FROM users
----

[NOTE]
====

> . When selecting the entire object, the query will always try to return the underlying object verbatim. For Java-serialized types, this means returning an underlying Java class instance, which can fail with a `ClassNotFoundException` if the class is not in the classpath of the client (or embedded server) JVM. To avoid this, you can select individual fields instead. This issue does not apply to Portable- or Compact-serialized types as sub-objects in these mappings and types are `GenericRecord` subclasses; `PortableGenericRecord` and `CompactGenericRecord` are present in the base distribution of Hazelcast.

Author: I disagree. My usage of commas is correct.

> . When selecting the entire object, the query will always try to return the underlying object verbatim. For Java-serialized types, this means returning an underlying Java class instance, which can fail with a `ClassNotFoundException` if the class is not in the classpath of the client (or embedded server) JVM. To avoid this, you can select individual fields instead. This issue does not apply to Portable- or Compact-serialized types as sub-objects in these mappings and types are `GenericRecord` subclasses; `PortableGenericRecord` and `CompactGenericRecord` are present in the base distribution of Hazelcast.

Reviewer 2: Keep the comma after types, then. I altered it because the sentence becomes very choppy when you use the other comma that I added. 

That would be the comma before "which", which is correct. If you prefer not to add the comma, change "which" to "that"

I was trying to do a really light-touch review this time, as you obviously take offence at every suggestion. However, if I was to have been picky, I would have pointed out that this sentence contains 41 words and needs to be re-written to address this.

> . When selecting the entire object, the query will always try to return the underlying object verbatim. For Java-serialized types, this means returning an underlying Java class instance, which can fail with a `ClassNotFoundException` if the class is not in the classpath of the client (or embedded server) JVM. To avoid this, you can select individual fields instead. This issue does not apply to Portable- or Compact-serialized types as sub-objects in these mappings and types are `GenericRecord` subclasses; `PortableGenericRecord` and `CompactGenericRecord` are present in the base distribution of Hazelcast.

Author: 90767538

. For Avro-serialized types, the returned objects are subclasses of `org.apache.avro.generic.GenericRecord`, whose (de)serialization is supported by Java clients only.
====

Using projections:
[source,sql]
----
SELECT (organization).id * 1000, ABS((organization).office.id) FROM users
----
Projections work as usual since field access expressions have virtually the same semantics and possible usage contexts as normal column projections.

==== Querying Cyclic Types

> Following examples use <<a-type, AType>>.

Author: 90767538

NOTE: When selecting the entire object, the query will always try to return the underlying object verbatim.
For Java Types, this means returning an underlying Java class instance, which can fail with a `ClassNotFoundException`
if the class is not in the classpath of the client (or embedded server) JVM.

> To avoid this, you can select a field-by-field instead. Additionally, this issue does not apply to Portable, Compact, or Avro types as sub-objects in these mappings and types are `GenericRecord` subclasses;

Reviewer: remeber that there are different clients that java. In particular will Avro be supported in them?

Author: fef8c8c2

INSERT and UPDATE queries are supported in a limited way, specifically:

- `INSERT` and `UPDATE` queries are disabled for mappings that reference cyclic UDTs anywhere in the type hierarchy.
- `INSERT` queries require specifying the full list of columns even if the column of a nested type needs to be set to `NULL`.
- `UPDATE` queries only work on the root column and also require the full list of columns and sub-columns to work.
Updating sub-columns is technically possible by specifying column projections in place of sub-columns that shouldn't be changed.
- Both `UPDATE` and `INSERT` queries use the Row Value expression, which is similar to the `VALUES` clause of an `INSERT` query.

=== Examples [[upsertExamples]]

> Following examples use <<users-mapping, `users` mapping>>, and <<organization-office-types, `Organization` and `Office` types>>.

Author: 90767538

UPDATE users SET organization = ((organization).id, (organization).name, ((organization).office.id, 'new-office-name'))
----

NOTE: Updating UDT columns requires providing a value for every column in the UDT and its child UDTs unless it needs to be set to `null`. Not providing full list of columns will cause a query validation error.

> Inserting with query parameter (Java only):

Reviewer: is it really java only? the example maybe (uses `Serializable`), but this could work in principle for portable and compact from other languages as well.

Author: I removed the "Java only" notice and added a note for Avro support on non-Java clients. (fef8c8c2)

UPDATE users SET organization = ((organization).id, (organization).name, ((organization).office.id, 'new-office-name'))
----

> NOTE: Updating UDT columns requires providing a value for every column in the UDT and its child UDTs unless it needs to be set to `null`. Not providing full list of columns will cause a query validation error.

Reviewer 2: ```suggestion
NOTE: To update UDT columns, a value must be provided for every column in the UDT and its child UDTs unless it needs to be set to `null`. If a full list of columns is not provided, a query validation error occurs.
```

Author: "A full list" sounds very strange.

Reviewer 2: Does a "partial list" sound strange?

Also, I didn't change that, that was from your original text

Reviewer 2: Oh, hang on, reading that again, it is the use of the indefinite article that you are questioning. 

Nah, you need it, whether it sounds strange to you or not.

Author: 90767538

hz.getSql().execute("INSERT INTO users VALUES (1, 'user1', ?)", organization);
----

> NOTE: For Avro-serialized types, the query parameters need to be subclasses of `org.apache.avro.generic.GenericRecord`, whose (de)serialization is supported by Java clients only.

Reviewer 2: ```suggestion
NOTE: For Avro-serialized types, the query parameters must be subclasses of `org.apache.avro.generic.GenericRecord`, whose (de)serialization is supported by Java clients only.
```

Author: 90767538

hz.getSql().execute("UPDATE users SET organization = ?", organization);
----

Updating nested UDT column with query parameter:

[source,java]
----
hz.getSql().execute("UPDATE users SET organization = ((organization).id, (organization).name, ?)", office);
----

> == Rolling Upgrade Notes

Reviewer: this affects also normal upgrade with persistent SQL catalog. both are EE features - we should have "enterprise" marker here

Author: fef8c8c2

hz.getSql().execute("UPDATE users SET organization = ?", organization);
----

Updating nested UDT column with query parameter:

[source,java]
----
hz.getSql().execute("UPDATE users SET organization = ((organization).id, (organization).name, ?)", office);
----

== Upgrade Notes
[.enterprise]*Enterprise*

> When performing a (normal or rolling) upgrade from version 5.3 to 5.4, you must drop all user-defined types and mappings with UDTs before the upgrade, and recreate them with the new semantics after upgrading.

Reviewer 2: ```suggestion
When performing a normal or rolling upgrade from version 5.3 to 5.4, you must drop all user-defined types and mappings with UDTs before the upgrade, and recreate them with the new semantics after upgrading.
```

[source,java]
----
hz.getSql().execute("UPDATE users SET organization = ((organization).id, (organization).name, ?)", office);
----

== Rolling Upgrade Notes

> You must drop all user-defined types and mappings with UDTs before the rolling upgrade, and recreate them with the new semantics after upgrading.

Reviewer: add information about versions to which this restricion applies. This documentation will probably not be updated in 5.5 and later

Author: 099f756d
